### ISSUE TITLE
[no-unused-vars] False positive for destructured function parameters

### ISSUE THREAD

**Issue by delucis** [typescript-eslint/typescript-eslint#2904]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": ["warn", {
      "args": "after-used",
      "ignoreRestSiblings": true
    }]
  }
}
```

```TS
const obj = { foo: 'bar' };
Object.entries(obj)
  .filter(([key, value]) => value === 'bar')
  .map(([key]) => key);
```

**Expected Result**

Although `key` is unused in the filter function, because it precedes a used argument, it should not raise a warning. I guess that `after-used` applies strictly to the root arguments rather than destructured array arguments, but perhaps this is a logical extension?

(Obviously there are other ways to write something like this, but they all seem likely to be less clean.)

**Actual Result**

ESLint raises `'key' is defined but never used. (@typescript-eslint/no-unused-vars)`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.11.0` |
| `@typescript-eslint/parser`        | `4.11.0` |
| `TypeScript`                       | `3.8.2` |
| `ESLint`                           | `7.16.0` |
| `node`                             | `10.20.1` |

I also tried forking this repo and adding the repro to `packages/eslint-plugin/tests/rules/no-unused-vars/no-unused-vars.test.ts` with the same results.

Adding

```ts
    {
      code: `
const obj = { foo: 'bar' };
Object.entries(obj)
  .filter(([key, value]) => value === 'bar')
  .map(([key]) => key);
      `,
      options: [{ args: 'after-used', ignoreRestSiblings: true }]
    }
```

to the array of `valid` test cases produced the following test error:

```
 FAIL  tests/rules/no-unused-vars/no-unused-vars.test.ts (9.246 s)
  ● no-unused-vars › valid › 
const obj = { foo: 'bar' };
Object.entries(obj)
  .filter(([key, value]) => value === 'bar')
  .map(([key]) => key);
      

    assert.strictEqual(received, expected)

    Expected value to strictly be equal to:
      0
    Received:
      1
    
    Message:
      Should have no errors but had 1: [ { ruleId: 'no-unused-vars',
        severity: 1,
        message: '\'key\' is defined but never used.',
        line: 4,
        column: 13,
        nodeType: 'Identifier',
        messageId: 'unusedVar',
        endLine: 4,
        endColumn: 16 } ]

      at testValidTemplate (../../node_modules/eslint/lib/rule-tester/rule-tester.js:608:20)
      at Object.RuleTester.it (../../node_modules/eslint/lib/rule-tester/rule-tester.js:885:25)
```

**Comment by bradzacher** [comment]
this is working as intended.
in an array destructuring, you can simply omit the unused elements.

```diff
  const obj = { foo: 'bar' };
  Object.entries(obj)
-   .filter(([key, value]) => value === 'bar')
+   .filter(([, value]) => value === 'bar')
    .map(([key]) => key);
```
